Bug #80458 PDOStatement::fetchAll() throws for upsert queries

kamil-tekiela opened the pull request:
While the previous fixes did improve error reporting, it turns out that a lack of a result set is not an error. MySQLnd always triggers an OOS error if you try to fetch from a prepared statement and there is no result. The behaviour is now more or less aligned with emulated prepares which would throw an exception with "General error" if a result set was not present. However, in my opinion, a lack of a result is not an error. For UPDATE/DELETE and any other query that doesn't produce results, it is expected that there will be no result returned. As can be observed by the 2 bug reports the community expects an empty array instead of some confusing exception.

This PR aims to fix the problem. First, we check in mysqlnd for a missing result. If there is no result then we return false without an exception. Second, when using emulated prepares and performing an UPSERT query we do the same. This aligns emulated prepares and native prepares back again and produces an empty array when calling fetchAll().

mvorisek commented on ext/pdo_mysql/tests/bug80458.phpt:
```diff
@@ -0,0 +1,83 @@
+    [0]=>
+    string(1) "6"
+  }
+}
```
fix EOL

nikic commented:
I'm not sure this fix is right. When linking against libmysqlclient and running the added test case, I get:

> Uncaught PDOException: SQLSTATE[HY000]: General error: 2053  in /home/nikic/php/php-src-libmysql/ext/pdo_mysql/tests/bug80458.php:11

This shows that libmysqlclient agrees with that this is an error. As such it needs to stay an error at least at the level of mysqlnd.

If we want to treat this as not an error in PDO, then I believe the fix needs to be on the side of PDO, not mysqlnd.

kamil-tekiela commented:
@nikic Ok, I agree. I reverted the change back and added a check in PDO to see if the statement produces a result or not. I have also added a more comprehensive test case. 
Let me know if this alternative solution is better.

nikic commented on ext/pdo_mysql/mysql_statement.c:
```diff
@@ -619,6 +619,17 @@ static int pdo_mysql_stmt_param_hook(pdo_stmt_t *stmt, struct pdo_bound_param_da
+
+	if(S->stmt) {
+		MYSQL_RES     *prepare_meta_result;
+		if(!(prepare_meta_result = mysql_stmt_result_metadata(S->stmt))) {
```
I don't think this is a function that should be called on each fetch. Shouldn't the result metadata already be stored inside `S->result`?

kamil-tekiela replied:
Makes sense. I replaced it with an existence check on `S->stmt->data->result` which I assume is allowed.

nikic commented on ext/pdo_mysql/tests/bug80458.phpt:
```diff
@@ -0,0 +1,156 @@
+$stmt5 = $db->prepare('CALL ret()');
+$stmt5->execute();
+var_dump($stmt5->fetchAll());
+$stmt5->nextRowset(); // needed to fetch the empty result set of CALL
```
Should this also do a fetchAll() after the nextRowset()? If I'm understanding the interpretation we're going with right, then that should also work.

kamil-tekiela replied:
Yes, it should. Added.

kamil-tekiela commented:
Not to forget. This PR is aiming to fix https://bugs.php.net/bug.php?id=80267 and https://bugs.php.net/bug.php?id=80458

nikic commented on ext/pdo_mysql/mysql_statement.c:
```diff
@@ -619,6 +619,11 @@ static int pdo_mysql_stmt_param_hook(pdo_stmt_t *stmt, struct pdo_bound_param_da
 {
 	pdo_mysql_stmt *S = (pdo_mysql_stmt*)stmt->driver_data;
+
+	if(S->stmt && S->stmt->data && !S->stmt->data->result) {
```
Why is this not just `if (!S->result)`?

In which case the same check below could also be dropped.

kamil-tekiela replied:
That's because I thought that with unbuffered queries `S->result` would be NULL but it looks like I was wrong. IIUC `S->result` is populated always with the metadata, correct? I changed it now.

ilmiont commented:
Had a production application break because a new container was created by CI pipeline on PHP 7.4.13 instead of PHP 7.4.12.

The application utilised a data access layer with a generic query method which would always try and fetch data from the statement, irrespective of the query type.

Our CI pipelines are allowed to build containers using the latest PHP patch version because on the general assumption that they should be safe to use. This was not the case and it has caused downtime for customers; I am greatly looking forward to 7.4.14 hopefully incorporating this PR and addressing 80267/80458.

kamil-tekiela commented:
> Had a production application break because a new container was created by CI pipeline on PHP 7.4.13 instead of PHP 7.4.12.
> 
> The application utilised a data access layer with a generic query method which would always try and fetch data from the statement, irrespective of the query type.
> 
> Our CI pipelines are allowed to build containers using the latest PHP patch version on the general assumption that it should be safe to use. This was not the case and it has caused downtime for customers; I am greatly looking forward to 7.4.14 hopefully incorporating this PR and addressing 80267/80458.

Thanks for the feedback. I assume you believe this behaviour is a bug rather than the expected behaviour. Does it mean your application is exclusively using native prepared statements? I also believe this is a bug, but emulated statements have triggered an exception in this scenario for years. This PR is trying to silence the exception in both emulated and native prepared statements.

nikic commented:
Confirming that this works on libmysqlclient as well. (The test fails due to type differences, but that's a different problem.)

nikic reviewed (approved):
Okay, this looks good to me now. I think this is the right direction. While performing fetches for upsert is incorrect if we're being pedantic, I don't see much benefit in enforcing that. Consistently being able to use fetch even if there are no results is more valuable.

nikic commented:
Merged as https://github.com/php/php-src/commit/a83cc03c138b8cf27a840bd7cd913eb7050e55ba in PHP-7.4 upwards.

ilmiont commented:
> > Had a production application break because a new container was created by CI pipeline on PHP 7.4.13 instead of PHP 7.4.12.
> > The application utilised a data access layer with a generic query method which would always try and fetch data from the statement, irrespective of the query type.
> > Our CI pipelines are allowed to build containers using the latest PHP patch version on the general assumption that it should be safe to use. This was not the case and it has caused downtime for customers; I am greatly looking forward to 7.4.14 hopefully incorporating this PR and addressing 80267/80458.
> 
> Thanks for the feedback. I assume you believe this behaviour is a bug rather than the expected behaviour. Does it mean your application is exclusively using native prepared statements? I also believe this is a bug, but emulated statements have triggered an exception in this scenario for years. This PR is trying to silence the exception in both emulated and native prepared statements.

Yes, we exclusively use native prepares.

I really don't want to get into the discussion of whether this was strictly-speaking a bug or not; to me as a PHP user, it most certainly was a bug because code that worked flawlessly in 7.4.12 broke overnight in 7.4.13. That defies basic expectations around semantic versioning. Glad to see this PR.
